tokenclient/tokenFlows.go

```diff
@@ -0,0 +1,196 @@
+// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Cloud Security Client Go contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+package tokenclient
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"github.com/sap/cloud-security-client-go/auth"
+	"github.com/sap/cloud-security-client-go/env"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+type Options struct {
+	HTTPClient *http.Client // TODO Default: basic http.Client with a timeout of 15 seconds
+	TLSConfig  *tls.Config  // TODO Default:
+}
+
+// RequestOptions allows to configure the token request
+type RequestOptions struct {
+	// Context carries the request context like the deadline or other values that should be shared across API boundaries.
+	Context context.Context
+	// Request parameters that shall be overwritten or added to the payload
+	Params map[string]string
+}
+
+type TokenFlows struct {
+	identity *env.Identity
+	options  Options
+	tokenURI string
+}
+
+type ClientError struct {
```

Comment:
What's the idea behind ClientError? Do you think using
`fmt.Errorf("something went wrong: %v", err)`
would work as well?

Reply:
Addition: Please use `%w` (instead of `%v`) so that errors can be unwrapped. 
Further reading: https://go.dev/blog/go1.13-errors

Reply:
@jkbschmid The original idea was in future to add more information, like status code into it, that can be parsed by the application in order to repeat that request, e.g. in case of `429` status code. So for the ClientCredentials request i would like to keep it and improve it...

Furthermore, I thought it would be a better design to have only one error type and also for wrapping causing errors. But i like more to use standard error `%w` for the TokenFlow initialization.

Reply:
for now I've removed ClientError, applications need to parse status code from error message.
"failed with status code '401' and payload: 'unauthenticated client'"

--> any better idea is appreciated!

Reply:
Thanks for the explanation! 
> can be parsed by the application


Then a custom error type is exactly the right choice, in my opinion :) 
We could create the initial errors the way you do it now via `ClientError{...}` (or even use separate error types for 4xx and 5xx) and if we need to compose errors (as Felix mentioned) using `fmt.Errorf("some more info: %w", clientErr)`.

This way the error becomes part of the API and consumers can react to different errors, e.g. via 
```go 
if errors.As(err, &clientError) &&  clientErr.StatusCode == 500{
...
}
```

Reply:
@jkbschmid  💯 cool - now it looks pretty!